src/protocol/libp2p/kademlia/mod.rs

```diff
@@ -395,12 +414,13 @@ impl Kademlia {
             .await;
 
         for info in peers {
-            self.service.add_known_address(&info.peer, info.addresses.iter().cloned());
+            let addresses = info.addresses();
+            self.service.add_known_address(&info.peer, addresses.clone().into_iter());
 
             if std::matches!(self.update_mode, RoutingTableUpdateMode::Automatic) {
                 self.routing_table.add_known_peer(
                     info.peer,
-                    info.addresses.clone(),
+                    addresses,
                     self.peers
                         .get(&info.peer)
                         .map_or(ConnectionType::NotConnected, |_| ConnectionType::Connected),
```

Comment:
This should go into `on_connection_established` and be completely removed from here. Instead of decreasing the scores of addresses we fail to dial with an objective to remove them later better to only add addresses we were able to reach from the very beginning.